### ISSUE TITLE
0.6-SNAPSHOT build failed:  Could not resolve all dependencies for configuration

### ISSUE THREAD

**Issue by ntarocco** [vanniktech/gradle-android-junit-jacoco-plugin#57]
I am trying to use the version 0.6.0-SNAPSHOT because I need your addition of 
`includeNoLocationClasses = true` 
parameter added with the fix to #51, but when I build the app with this version of your plugin I have:

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':app'.
> Failed to notify project evaluation listener.
   > Could not resolve all dependencies for configuration ':app:_storeDebugApk'.
      > A problem occurred configuring project ':nspcoreui'.
         > Android tasks have already been created.
This happens when calling android.applicationVariants,
android.libraryVariants or android.testVariants.
Once these methods are called, it is not possible to
continue configuring the model.
   > Android tasks have already been created.
This happens when calling android.applicationVariants,
android.libraryVariants or android.testVariants.
Once these methods are called, it is not possible to
continue configuring the model.

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
```

See attached full stack trace for more details.
Everything works instead with the release 0.5.
Can you help me out? I need it urgently if possible...

[stacktrace.txt](https://github.com/vanniktech/gradle-android-junit-jacoco-plugin/files/817314/stacktrace.txt)

The build.grade of mymodule1 is:

```

apply plugin: 'com.android.library'
apply plugin: 'me.tatarka.retrolambda'
apply plugin: 'com.vanniktech.android.junit.jacoco'

apply from: 'linter.gradle'

junitJacoco {
    jacocoVersion = rootProject.ext.jacocoVersion // type String
//    ignoreProjects = [] // type String array
    excludes =  rootProject.ext.jacocoExcludes
//    includeNoLocationClasses = true
}

android {
    compileSdkVersion rootProject.ext.compileSdkVersion
    buildToolsVersion rootProject.ext.buildToolsVersion

    dataBinding {
        enabled = true
    }

    defaultConfig {
        minSdkVersion 15
        versionCode 1
        versionName "1.0"
    }
    lintOptions {
        abortOnError false
    }
    buildTypes {
        debug {
            debuggable true
            testCoverageEnabled = true
        }
        release {
            testCoverageEnabled = false
        }
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
}

dependencies {
    compile fileTree(dir: 'libs', include: ['*.jar'])
    compile "io.reactivex:rxjava:${rootProject.ext.rxJavaVersion}"
    compile "com.jakewharton.timber:timber:${rootProject.ext.timberVersion}"

    testCompile 'junit:junit:4.12'
    testCompile 'org.mockito:mockito-core:1.10.19'

    compile "com.android.support:support-annotations:${rootProject.ext.supportLibVersion}"

    compile "com.android.support:design:${rootProject.ext.supportLibVersion}"
    compile "com.android.support:appcompat-v7:${rootProject.ext.supportLibVersion}"
    compile "com.android.support:support-v4:${rootProject.ext.supportLibVersion}"

    compile "com.android.support:cardview-v7:${rootProject.ext.supportLibVersion}"
    compile "com.android.support:recyclerview-v7:${rootProject.ext.supportLibVersion}"
}

```

**Comment by vanniktech** [comment]
Version should be `0.6.0-SNAPSHOT` and not `0.6-SNAPSHOT`

**Comment by ntarocco** [comment]
Yes it is, I am using: 
`classpath 'com.vanniktech:gradle-android-junit-jacoco-plugin:0.6.0-SNAPSHOT'`
sorry typo...
I have also tried to compile your project to be sure, but same issue...

**Comment by vanniktech** [comment]
The plugin should be applied on `root/build.gradle`. Is it in your case?

**Comment by ntarocco** [comment]
Yes.
I repeat: using version 0.5.0 everything works perfectly.
If I put 0.6.0-SNAPSHOT with NO other changes, I have that compilation error.

**Comment by vanniktech** [comment]
Oh okay. I'll have a look sometime soon

**Comment by ntarocco** [comment]
I have an update. I checked your code, and the problem is that you are doing this:

         ```
subProject.android {
                testOptions {
                    unitTests.all {
                        jacoco {
                            includeNoLocationClasses = true
                        }
                    }
                }
            }
```

In my app, I have already this defined:

```
testOptions {
        unitTests.all {
            maxHeapSize = "1512m"
        }
    }
```

So the build fails because there is already the same object already created, why you should try to "merge" it.
Does it help?

**Comment by vanniktech** [comment]
Hmm interesting. This new flag actually came from @f3ist. Do you guys have any idea how we can set the flag without destroying the consumers configuration?

**Comment by ntarocco** [comment]
I am trying right now but I am not an expert, so I didn't find a solution yet.

**Comment by ntarocco** [comment]
@vanniktech I have found a solution, I will create a PR soon

**Comment by vanniktech** [comment]
Cool looking forward to it

**Comment by ntarocco** [comment]
I have created the PR to fix it. Do you plan to release the 0.6.0 soon then?

**Comment by vanniktech** [comment]
If it fixes the issue then yes.

**Comment by vanniktech** [comment]
@ntarocco `0.6.0-SNAPSHOT` should be deployed soon. I'd give it about half an hour. Can you then just verify it for yourself again? If it's working for you I'll release 0.6.0 tonight latest tomorrow.

**Comment by armatys** [comment]
Hi,
I was using the `0.6.0-SNAPSHOT` version with `includeNoLocationClasses = true` and 2 days ago (around the time commit 1cf057c14dd80a1b49aa81ea9c1e5f7e1d7ade9e has been merged), jacoco stopped counting unit tests that are using robolectric (which require the use of `includeNoLocationClasses`). So in effect, it seems that setting `includeNoLocationClasses` extension has no effect anymore.

@ntarocco As I cannot reproduce your error in my project, maybe you could verify if this change would work for you:
```
subProject.android.testOptions.unitTests.all {
    it.jacoco.includeNoLocationClasses = extension.includeNoLocationClasses
}
```
instead of this:
```
subProject.android.testOptions.unitTests {
    it.testTasks.each {
        it.jacoco.includeNoLocationClasses = extension.includeNoLocationClasses
    }
}
```

From what I see, property called `testTasks` is private - I don't see it in [TestOptions.UnitTestOptions](https://google.github.io/android-gradle-dsl/current/com.android.build.gradle.internal.dsl.TestOptions.UnitTestOptions.html) documentation.

**Comment by ntarocco** [comment]
@armatys from my tests, if you use

`
subProject.android.testOptions.unitTests.all {
    it.jacoco.includeNoLocationClasses = extension.includeNoLocationClasses
}
`

then you will override any configuration already declared in the your Android project.
The .all() method redefine the config, if I am not mistaken.
I agree that in the doc you don't see it, but I went into the code to find that. See:
https://android.googlesource.com/platform/tools/base/+/d71bf1b/build-system/gradle-core/src/main/groovy/com/android/build/gradle/internal/dsl/TestOptions.groovy

**Comment by armatys** [comment]
Ok, I see.

I did some "println" debugging, and it turned out that when using this code
```
subProject.android.testOptions.unitTests {
    println("it.testTasks: ${it.testTasks}")
    it.testTasks.each {
        it.jacoco.includeNoLocationClasses = extension.includeNoLocationClasses
    }
}
```
`it.testTasks` is always an empty array, and in the end `includeNoLocationClasses` is not set at all. Could you check as well?

**Comment by ntarocco** [comment]
@armatys I have just made a test.
My project has a main app with 2 submodules.

Code of this plugin:

    subProject.android.testOptions.unitTests {
		println("*********************************************************************************************")
		println("it: ${it}")
		println("it.testTasks: ${it.testTasks}")
		println("it.returnDefaultValues: ${it.returnDefaultValues}")

		it.testTasks.each {
			println("----------------------------------------------------")

			it.jacoco.includeNoLocationClasses = extension.includeNoLocationClasses

			println("it: ${it}")
			println("it.jacoco: ${it.jacoco}")
			println("it.jacoco.includeNoLocationClasses: ${it.jacoco.includeNoLocationClasses}")
			println("----------------------------------------------------")
		}
		println("*********************************************************************************************")
	}
	
When I build, console outputs:

	Main App
	*********************************************************************************************
	it: com.android.build.gradle.internal.dsl.TestOptions$UnitTestOptions@7796f1ca
	it.testTasks: [task ':app:testFlavorDebugUnitTest', task ':app:testFlavorReleaseUnitTest']
	it.returnDefaultValues: true
	----------------------------------------------------
	it: task ':app:testFlavorDebugUnitTest'
	it.jacoco: org.gradle.testing.jacoco.plugins.JacocoTaskExtension_Decorated@30513349
	it.jacoco.includeNoLocationClasses: true
	----------------------------------------------------
	----------------------------------------------------
	it: task ':app:testFlavorReleaseUnitTest'
	it.jacoco: org.gradle.testing.jacoco.plugins.JacocoTaskExtension_Decorated@33a6f0fe
	it.jacoco.includeNoLocationClasses: true
	----------------------------------------------------
	*********************************************************************************************

	Submodule 1
	*********************************************************************************************
	it: com.android.build.gradle.internal.dsl.TestOptions$UnitTestOptions@45493a95
	it.testTasks: [task ':submodule1:testDebugUnitTest', task ':submodule1:testReleaseUnitTest']
	it.returnDefaultValues: false
	----------------------------------------------------
	it: task ':submodule1:testDebugUnitTest'
	it.jacoco: org.gradle.testing.jacoco.plugins.JacocoTaskExtension_Decorated@739d991c
	it.jacoco.includeNoLocationClasses: true
	----------------------------------------------------
	----------------------------------------------------
	it: task ':submodule1:testReleaseUnitTest'
	it.jacoco: org.gradle.testing.jacoco.plugins.JacocoTaskExtension_Decorated@77a1d9d8
	it.jacoco.includeNoLocationClasses: true
	----------------------------------------------------
	*********************************************************************************************

	Submodule 2
	*********************************************************************************************
	it: com.android.build.gradle.internal.dsl.TestOptions$UnitTestOptions@141b1481
	it.testTasks: [task ':submodule2:testDebugUnitTest', task ':submodule2:testReleaseUnitTest']
	it.returnDefaultValues: false
	----------------------------------------------------
	it: task ':submodule2:testDebugUnitTest'
	it.jacoco: org.gradle.testing.jacoco.plugins.JacocoTaskExtension_Decorated@834385d
	it.jacoco.includeNoLocationClasses: true
	----------------------------------------------------
	----------------------------------------------------
	it: task ':submodule2:testReleaseUnitTest'
	it.jacoco: org.gradle.testing.jacoco.plugins.JacocoTaskExtension_Decorated@3ed584ea
	it.jacoco.includeNoLocationClasses: true
	----------------------------------------------------
	*********************************************************************************************
	
The app and the submodules are set with includeNoLocationClasses to true.

	junitJacoco {
		jacocoVersion = "0.7.9"
		includeNoLocationClasses = true
		excludes= ['**/R.class','**/R$*.class', '**/*$$*']
	}

and the main app ONLY, on top of that, has returnDefaultValues set to true:

	testOptions {
        resultsDir = "$project.buildDir/output/test_results"
        unitTests.returnDefaultValues = true
        unitTests.all {
            maxHeapSize = "1512m"
        }
    }

As you can see, the includeNoLocationClasses is correctly set for the main app and the submodules, and the returnDefaultValues is set to true only for the main app.

**Comment by armatys** [comment]
Thanks @ntarocco I have a few more observations.

I tried your code, but `it.testTasks` still returned empty array.

Then, looking at your first post, it seems you are calling `apply plugin: 'com.vanniktech.android.junit.jacoco'` directly in your module:

```
my_project/
    build.grade <-- I applied here (top-level)
    settings.gradle
    app/
        build.gradle
        src/
    mymodule1/
        build.gradle <-- you applied here (module level)
        src/
```

Is that the case?
Because when I moved the plugin application to `mymodule1`, then I was able to reproduce your initial error. 

And when testing with the latest snapshot, only then `it.testTasks` was returning a non-empty array.

Last observation is that there is a difference between using the original code:

```
subProject.android {
            testOptions {
                unitTests.all {
                    jacoco {
                        includeNoLocationClasses = extension.includeNoLocationClasses
                    }
                }
            }
        }
```

which results in failure (`Android tasks have already been created`), while using this:

```
subProject.android.testOptions.unitTests.all {
    it.jacoco.includeNoLocationClasses = extension.includeNoLocationClasses
}
```

(even though it looks very similar) does not throw an error. And it also works when applying the plugin to the top-level `build.gradle` file.

**Comment by ntarocco** [comment]
@armatys You are right, I applied it like this:

    my_project/
        build.grade <-- just added classpath dependency
        settings.gradle
        app/
            build.gradle <-- I applied it here
            src/
        mymodule1/
            build.gradle <-- I applied it here
            src/
       mymodule2/
            build.gradle <-- I applied it here
            src/

And this cause the issue, while if I apply the plugin at the root level, it works.
But then: is it not a bit "weak" to oblige to apply it at the root level otherwise everything is broken?

@vanniktech applying the plugin at the root and not to a module works. Up to you to decide what to do.

**Comment by vanniktech** [comment]
In the Readme it says that the plugin should be applied to the root build.gradle. So I guess everything is working is intended right?

Which means that armatys used it incorrectly?

**Comment by ntarocco** [comment]
@vanniktech no it means that I have used it incorrectly in my project, and his code was right.
You should decline my PR and re-add back his code that you have apparently reverted.

**Comment by vanniktech** [comment]
I'm a bit confused now. Which code should be there now? I didn't revert any commits.

**Comment by vanniktech** [comment]
Ok back to it again. I merged the PR with the changes. @armatys @ntarocco could you please test it again with the latest SNAPSHOT? Should be available in ~30 minutes.

**Comment by armatys** [comment]
Tested, on my end works fine.

**Comment by vanniktech** [comment]
Alright cool. If no one complains further I'll try to release `0.6.0` tonight.

**Comment by ntarocco** [comment]
Just tested, works in my project!

**Comment by vanniktech** [comment]
Alright thanks for all of your work.

**Comment by vanniktech** [comment]
@ntarocco @armatys I just released `0.6.0` thanks once again :+1:
